marxs/optics/multiLayerMirror.py

```diff
@@ -0,0 +1,104 @@
+    '''Multilayer mirror with varying layer thickness along one axis
+    
+    The distance between layers changes along the local y axis.
+    All reflectivity data is assumed to be for a single, desired angle. There
+    is currently no way to enter varying reflection that depends on the angle
+    of incidence.
+    Provide reflectivity data in a file with columns:
+    	'X(mm)' - position along the "changing" axis
+    	'Peak lambda' - wavelength with maximum reflection at a given position
+    	'Peak' - maximum reflection at a given position
+    	'FWHM(nm)' - full width half max, measure of width of reflection Gaussian peaks
+    
+    Parameters
+    ----------
+    fileCode: string
+    	serial number of mirror, used to open reflection data file
+    '''
+    def __init__(self, fileCode, **kwargs):
+        self.fileName = './marxs/optics/docs/' + fileCode + '.txt'
+        if ('zoom' not in kwargs):
+        	kwargs['zoom'] = np.array([1, 24.5, 12])   # in mm
+        super(MultiLayerMirror, self).__init__(**kwargs)
+
+    def process_photons(self, photons):
+        # eliminate photons that do not reach the mirror
+        doesIntersect, intersection, local = self.intersect(photons['dir'], photons['pos'])
+        photons['probability'] *= doesIntersect
+        
+        # split polarization into parallel and other components
+        beam_dir = (photons['dir'][:,0:3]).copy()
+        for i in range(0, len(beam_dir)):
+        	beam_dir[i] /= np.linalg.norm(beam_dir[i])
+        # v_1 is parallel to plane, v_2 is third direction remaining
+        v_1 = np.cross(beam_dir, self.geometry['e_x'][0:3])
+        v_2 = np.cross(beam_dir, v_1)
+        # find polarization in each direction
+        polarization = (photons['polarization'][:,0:3]).copy()
+        p_v_1 = np.zeros(len(polarization))
+        p_v_2 = np.zeros(len(polarization))
+        for i in range(0, len(polarization)):
+        	v_1[i] /= np.linalg.norm(v_1[i])
+        	v_2[i] /= np.linalg.norm(v_2[i])
+        	p_v_1[i] = np.dot(polarization[i], v_1[i])   # just the cosine between the vectors
+        	p_v_2[i] = np.dot(polarization[i], v_2[i])
+        
+        # reflect (change direction)
+        directions = photons['dir']
+        pos4d_inv = np.linalg.inv(self.pos4d)
+        directions = directions.T
+        directions = np.dot(pos4d_inv, directions)
+        directions[0,:] *= -1
+        directions = np.dot(self.pos4d, directions)
+        directions = directions.T
+        photons['dir'] = directions
+        
+        # adjust polarization vectors after reflection
+        # find new v_2
+        new_beam_dir = (photons['dir'][:,0:3]).copy()
+        for i in range(0, len(new_beam_dir)):
+        	new_beam_dir[i] /= np.linalg.norm(new_beam_dir[i])
+        new_v_2 = np.cross(new_beam_dir, v_1)
+        for i in range(0, len(polarization)):
+        	photons['polarization'][i,0:3] = p_v_1[i] * v_1[i] + p_v_2[i] * new_v_2[i]
```

Comment:
Some fancy indexing could avoid this for loop. Something like
`photons['polarization'] = p_v_ * v_1[np.newaxis, :] + p_v_2 * new_v_2[np.newaxis, :]`